homeassistant/components/climate/generic_thermostat.py

```diff
@@ -302,65 +390,143 @@ def _async_update_temp(self, state):
             _LOGGER.error('Unable to update from sensor: %s', ex)
 
     @callback
-    def _async_control_heating(self):
+    def _async_control_thermostat(self):
```

Comment:
Phew… I've tried to construct the state machine for this one in my head, and kind of failed. I wonder whether we can't cut back on complexity:

What use is `self._target_temp`? I think with heating and / or AC, two simple rules would be sufficient:
* If (and only if) you specify a `heater_control` entity, you can and must specify a `CONF_TARGET_TEMP_LOW`
* If (and only if) you specify a `ac_control` entity, can can and must specify a `CONF_TARGET_TEMP_HIGH`

The temperature for heating is always taken from `CONF_TARGET_TEMP_LOW`, the temperature for cooling is always taken from `CONF_TARGET_TEMP_HIGH`

This would allow us to get rid of:
* `self._target_temp`
* `self._dual_temp` (and therefore many `if self._dual_temp` checks)
* `CONF_COLD_TOLERANCE and CONF_HOT_TOLERANCE`

Reply:
Thank you! I think I was somehow stuck with keeping some style from the previous method. I'll refactor using your suggestions as it makes a lot more sense.

Reply:
CONF_COLD_TOLERANCE and CONF_HOT_TOLERANCE should stay because that allows you better control. You can start heating at -0.3 and stop it at +0.1 for example.